### ISSUE TITLE
`test_branch_creation_before_gc` is flaky

### ISSUE THREAD

**Issue by aome510** [neondatabase/neon#2140]
The latest failure can be found in https://github.com/neondatabase/neon/runs/7453848505?check_suite_focus=true.

Relevant `pageserver.log`:
```
2022-07-21T16:35:26.921862Z  INFO cfg failpoint: before-timeline-gc sleep(2000)
2022-07-21T16:35:26.962664Z  INFO check_checkpoint_distance 984415, layer size 1193835
2022-07-21T16:35:26.974389Z  INFO check_checkpoint_distance 917536, layer size 1111879
2022-07-21T16:35:27.315322Z  INFO compact{timeline=3924f4be79e894ef91831b64ffc5b1cf tenant=26aaa31ab51e4229bd97d8ea0a0d878c}: new image layer /tmp/test_output/test_branch_creation_before_gc/repo/tenants/26aaa31ab51e4229bd97d8ea0a0d878c/timelines/3924f4be79e894ef91831b64ffc5b1cf/000000000000000000000000000000000000-030000000000000000000000000000000002__00000000043AAA60.VyTitBSb.temp
2022-07-21T16:35:27.962652Z  INFO /timeline_create{tenant=26aaa31ab51e4229bd97d8ea0a0d878c new_timeline=None lsn=Some(0/2D469A8)}: branched timeline d024d8a92372ec29855acb42ab011ad0 from 3924f4be79e894ef91831b64ffc5b1cf at 0/2D469A8
2022-07-21T16:35:27.962775Z  INFO /timeline_create{tenant=26aaa31ab51e4229bd97d8ea0a0d878c new_timeline=None lsn=Some(0/2D469A8)}:loading local timeline: loaded layer map with 0 layers at 0/2D469A8
2022-07-21T16:35:28.002133Z  INFO wal_receiver_main: Processing timeline update: Add(ZTenantTimelineId { tenant_id: 26aaa31ab51e4229bd97d8ea0a0d878c, timeline_id: d024d8a92372ec29855acb42ab011ad0 })
2022-07-21T16:35:28.002170Z  INFO POST /v1/tenant/26aaa31ab51e4229bd97d8ea0a0d878c/timeline 201 Created
....
```

## My hypothesis

The `test_branch_creation_before_gc` test expects branch creation task to be failed because a `do_gc` request is processed before and makes the specified starting LSN invalid.

However, `do_gc` requires to read-lock the tenant state before
```rust
let repo = tenant_mgr::get_repository_for_tenant(tenantid)?; // -> read-lock here

let gc_horizon: u64 = caps
    .get(4)
    .map(|h| h.as_str().parse())
    .unwrap_or_else(|| Ok(repo.get_gc_horizon()))?;

// Use tenant's pitr setting
let pitr = repo.get_pitr_interval();
let result = repo.gc_iteration(Some(timelineid), gc_horizon, pitr, true)?;
...
```

and the write-lock can be acquired by the compaction thread:

```rust
let _layer_removal_cs = self.layer_removal_cs.lock().unwrap();

let target_file_size = self.get_checkpoint_distance();

// Define partitioning schema if needed
if let Ok(pgdir) =
    tenant_mgr::get_local_timeline_with_load(self.tenant_id, self.timeline_id) // write-lock is acquired here
{
...
```

In other words, if a compaction happens between, `do_gc` will be delayed. As a result, branch creation request is processed before the GC request, which makes the test failed.

I added a delay to somewhat prevent this behaviour, but that's not enough: https://github.com/neondatabase/neon/blob/main/test_runner/batch_others/test_branch_and_gc.py#L166-L169
